### PR TITLE
Extract proxy credential handling into reusable normalize_proxy_url utility

### PR DESCRIPTION
**Pull request by agentydragon**
## Summary
Refactors proxy credential handling to extract a reusable `normalize_proxy_url()` utility function that splits embedded credentials from proxy URLs into explicit `Proxy-Authorization` headers. This addresses a urllib3 v2 incompatibility where credentials embedded in proxy URLs are not automatically sent on HTTPS CONNECT tunnels.

## Key Changes

- **New `normalize_proxy_url()` utility** (`auth_proxy/vars.py`): Extracts username/password from proxy URLs and returns both a sanitized URL and a `Proxy-Authorization` header dict. This is required for urllib3 v2 compatibility when using raw HTTP clients (e.g., kubernetes Python client).

- **Refactored k8s_secrets.py**: Replaced inline proxy credential extraction logic with calls to `normalize_proxy_url()`, simplifying the code and removing a CLEANUP comment about persisting credentials to disk.

- **Enhanced DeferredOtlpExporter**: Now accepts a `requests.Session` parameter in `configure()` to support proxy and CA configuration from the caller's environment.

- **Updated session_start handler**: 
  - Added `caller_env` to `CallerContext` to preserve the original environment
  - New `_build_otlp_session()` helper to construct a requests.Session with proxy and CA settings
  - Falls back to caller's environment proxy when no local auth proxy exists (UDS mode)

- **New integration test** (`test_k8s_proxy_integration.py`): Verifies k8s secrets retrieval through an egress proxy in UDS mode (no local TCP auth proxy), ensuring `normalize_proxy_url()` correctly handles credential extraction for urllib3 v2.

- **New unit tests**: 
  - `test_vars.py`: Comprehensive tests for `get_proxy_url()` and `normalize_proxy_url()` with various credential formats
  - `test_tracing.py`: Tests for `DeferredOtlpExporter` configuration and buffering behavior

- **Utility enhancement** (`util/docker.py`): Added `get_docker_network_gateway()` helper for test infrastructure.

## Implementation Details

The `normalize_proxy_url()` function handles edge cases including:
- URLs with special characters in passwords (base64-encoded correctly)
- URLs with username but no password
- URLs without credentials (returned unchanged)
- Port preservation in the sanitized URL

The refactoring maintains backward compatibility while enabling proper proxy authentication in environments using raw urllib3 clients without a local TCP auth proxy.

https://claude.ai/code/session_01N3RaewRkrXRNzfSAF2PR8T

### CODE REVIEW

File: devinfra/claude/auth_proxy/vars.py [hunk 1]
```diff
@@ -19,12 +21,36 @@
...
     for var in PROXY_ENV_VARS:
-        if value := os.environ.get(var):
+        if value := env.get(var):
             return value
     return None
+
+
+def get_upstream_proxy_url() -> str | None:
+    """Get the upstream proxy URL from os.environ."""
+    return get_proxy_url(dict(os.environ))
```
Comment (agentydragon) [inline devinfra/claude/auth_proxy/vars.py:57]: This feels like a footgun. If we wanna be resilient against potential changes reading from environment like this should not be hidden. Because I'm not sure what happens if they ever rotate the proxy creds. Whether they will also kill our hook daemon or not. So we need to have resilience to caller env in hook coming with new creds

File: devinfra/claude/hook_daemon/session_start/BUILD.bazel [hunk 2]
```diff
@@ -139,18 +139,40 @@ py_library(
...
     ],
 )
 
 # === Tests ===
 
+py_test(
+    name = "test_k8s_proxy_integration",
+    srcs = ["test_k8s_proxy_integration.py"],
+    imports = ["../../../.."],
+    local = True,
```
Comment (agentydragon) [inline devinfra/claude/hook_daemon/session_start/BUILD.bazel]: Do not write local tests. This has no reason to not be runnable in rbe

File: devinfra/claude/hook_daemon/session_start/handler.py [hunk 3]
```diff
@@ -418,7 +437,7 @@ async def run_session(
                     session_dir=paths.session_dir,
                     combined_ca_path=combined_ca,
                     config=hook_config,
-                    proxy=setup.auth_proxy.proxy_url if setup.auth_proxy else None,
+                    proxy=(setup.auth_proxy.proxy_url if setup.auth_proxy else None) or get_proxy_url(ctx.caller_env),
```
Comment (agentydragon) [inline devinfra/claude/hook_daemon/session_start/handler.py]: This looks like something you could simplify

File: devinfra/claude/hook_daemon/session_start/test_k8s_proxy_integration.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+from devinfra.claude.testing.proxy_ca import generate_server_cert
+from util.docker import get_docker_network_gateway
+from util.net import pick_free_port, wait_for_port
+
+pytest_plugins = ["devinfra.claude.testing.mitmproxy_fixture"]
+
+_FAKE_SECRETS: dict[str, dict[str, str]] = {"github-token": {"token": "fake-github-token"}}
+
+
+class _K8sHandler(http.server.BaseHTTPRequestHandler):
```
Comment (agentydragon) [inline devinfra/claude/hook_daemon/session_start/test_k8s_proxy_integration.py:35]: Wondering if this could be made easier if using some higher level http server api

File: devinfra/claude/hook_daemon/session_start/test_k8s_proxy_integration.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+def mock_k8s_server(tmp_path: Path, proxy_net: docker.models.networks.Network) -> Generator[MockK8sServer]:
+    """HTTPS mock k8s API server on the host, reachable from mitmproxy via the bridge gateway."""
+    gateway = get_docker_network_gateway(proxy_net)
+    port = pick_free_port(host="0.0.0.0")
+
+    cert_pem, key_pem = generate_server_cert(gateway)
+    cert_file = tmp_path / "server.crt"
+    key_file = tmp_path / "server.key"
+    cert_file.write_bytes(cert_pem)
+    key_file.write_bytes(key_pem)
```
Comment (agentydragon) [inline devinfra/claude/hook_daemon/session_start/test_k8s_proxy_integration.py:71]: You might as well make the writing happen in the fixture

File: devinfra/claude/hook_daemon/session_start/test_k8s_proxy_integration.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+    server = http.server.HTTPServer(("0.0.0.0", port), _K8sHandler)
+    server.socket = tls_ctx.wrap_socket(server.socket, server_side=True)
+
+    thread = threading.Thread(target=server.serve_forever, daemon=True)
+    thread.start()
+    wait_for_port("127.0.0.1", port, timeout_secs=5)
+
+    yield MockK8sServer(url=f"https://{gateway}:{port}")
+    server.shutdown()
```
Comment (agentydragon) [inline devinfra/claude/hook_daemon/session_start/test_k8s_proxy_integration.py]: Finally?

File: devinfra/claude/hook_daemon/test_tracing.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+class _FailingAdapter(requests.adapters.HTTPAdapter):
+    def send(self, *args, **kwargs):
+        raise requests.exceptions.ConnectionError("unreachable")
+
+
+@pytest.fixture
+def session() -> requests.Session:
+    s = requests.Session()
+    s.mount("https://", _FailingAdapter())
```
Comment (agentydragon) [inline devinfra/claude/hook_daemon/test_tracing.py:24]: Maybe httpx or such might make the mocking nicer

File: devinfra/claude/hook_daemon/test_tracing.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    exporter = DeferredOtlpExporter()
+    config = OtelConfig(endpoint=_ENDPOINT)
+    exporter.configure(config, session=session)
+    inner = exporter._inner
+    exporter.configure(config, session=session)
+    assert exporter._inner is inner
+
+
+def test_configure_no_endpoint_is_noop() -> None:
+    exporter = DeferredOtlpExporter()
```
Comment (agentydragon) [inline devinfra/claude/hook_daemon/test_tracing.py:55]: Fixture?

File: devinfra/claude/hook_daemon/test_tracing.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+def session() -> requests.Session:
+    s = requests.Session()
+    s.mount("https://", _FailingAdapter())
+    return s
+
+
+def test_configure_does_not_crash_on_export_failure(session: requests.Session) -> None:
+    """configure() logs a warning and continues if the initial span flush fails."""
+    exporter = DeferredOtlpExporter()
+    exporter.export([ReadableSpan(name="test-span")])
```
Comment (agentydragon) [inline devinfra/claude/hook_daemon/test_tracing.py]: Span fixtures?

File: devinfra/claude/auth_proxy/vars.py (hunk 1 above)
Comment (agentydragon) [reply devinfra/claude/auth_proxy/vars.py:57]: Fixed. Removed the `get_upstream_proxy_url()` fallback from `k8s_secrets.py` — callers now must supply `proxy` explicitly. In `handler.py`, extracted `proxy_url` as a variable shared by both k8s and OTLP: `(setup.auth_proxy.proxy_url if setup.auth_proxy else None) or get_proxy_url(ctx.caller_env)`. OTLP now routes through the TCP auth proxy when available, so credential rotation is handled by the existing bazel-wrapper RPC mechanism — `caller_env` is only used as fallback when no local proxy exists.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: devinfra/claude/hook_daemon/session_start/BUILD.bazel (hunk 2 above)
Comment (agentydragon) [reply devinfra/claude/hook_daemon/session_start/BUILD.bazel]: Fixed. Removed `local = True`, added `tags = ["e2e", "external"]` to match the `container_e2e` test pattern.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: devinfra/claude/hook_daemon/session_start/handler.py (hunk 3 above)
Comment (agentydragon) [reply devinfra/claude/hook_daemon/session_start/handler.py]: Fixed. Extracted `proxy_url` before the k8s call; both k8s and OTLP now use it directly.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: devinfra/claude/hook_daemon/session_start/test_k8s_proxy_integration.py (hunk 4 above)
Comment (agentydragon) [reply devinfra/claude/hook_daemon/session_start/test_k8s_proxy_integration.py:35]: Left as `http.server` — the test needs a real TCP server reachable from the mitmproxy Docker container, so a mock-request library won't work. `fastapi`+`uvicorn` are available if you'd prefer, but I figured the one-route handler didn't justify the added complexity.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: devinfra/claude/hook_daemon/session_start/test_k8s_proxy_integration.py (hunk 5 above)
Comment (agentydragon) [reply devinfra/claude/hook_daemon/session_start/test_k8s_proxy_integration.py:71]: Fixed. Extracted a `ca_file` fixture that writes `mitmproxy_proxy.ca_cert_pem` to `tmp_path`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: devinfra/claude/hook_daemon/session_start/test_k8s_proxy_integration.py (hunk 6 above)
Comment (agentydragon) [reply devinfra/claude/hook_daemon/session_start/test_k8s_proxy_integration.py]: Fixed. Wrapped the `yield` in `try/finally`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: devinfra/claude/hook_daemon/test_tracing.py (hunk 7 above)
Comment (agentydragon) [reply devinfra/claude/hook_daemon/test_tracing.py:24]: `OTLPSpanExporter` accepts a `requests.Session`, so staying in requests-land is necessary here. `_FailingAdapter` is a one-liner and tests exactly what we care about. Left as-is.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: devinfra/claude/hook_daemon/test_tracing.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
```
[the remark this answers is not in the document]
Comment (agentydragon) [reply devinfra/claude/hook_daemon/test_tracing.py:37]: `ReadableSpan` is a concrete class in opentelemetry-sdk 1.x with all-defaulted constructor parameters — `ReadableSpan(name="test-span")` constructs fine. Left as-is.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: devinfra/claude/hook_daemon/test_tracing.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+def test_configure_no_endpoint_is_noop() -> None:
+    exporter = DeferredOtlpExporter()
+    exporter.configure(OtelConfig(endpoint=None), session=requests.Session())
+    assert exporter._inner is None
+
+
+def test_export_buffers_before_configure() -> None:
+    exporter = DeferredOtlpExporter()
+    span = ReadableSpan(name="test-span")
+    assert exporter.export([span]) == SpanExportResult.SUCCESS
```
[the remark this answers is not in the document]
Comment (agentydragon) [reply devinfra/claude/hook_daemon/test_tracing.py:62]: Same as above — `ReadableSpan(name=...)` is constructible. Left as-is.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: util/docker.py [hunk 12]
```diff
@@ -3,6 +3,22 @@
 from __future__ import annotations
 
 import aiodocker
+import docker.models.networks
+
+
```
[the remark this answers is not in the document]
Comment (agentydragon) [reply util/docker.py:8]: Fixed. Added `@pypi//docker` to `//util:docker` deps in `util/BUILD.bazel`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: devinfra/claude/auth_proxy/vars.py [hunk 13]
```diff
@@ -19,12 +21,36 @@
...
+    Returns (url_without_credentials, proxy_headers_dict).
+    If the URL has no credentials, returns (url, {}).
     """
+    parsed = urlparse(proxy_url)
+    if not parsed.username:
+        return proxy_url, {}
+    password = parsed.password or ""
+    auth = base64.b64encode(f"{parsed.username}:{password}".encode()).decode()
+    proxy_headers = {"Proxy-Authorization": f"Basic {auth}"}
+    hostname = parsed.hostname or ""
```
[the remark this answers is not in the document]
Comment (agentydragon) [reply devinfra/claude/auth_proxy/vars.py]: Fixed. `normalize_proxy_url()` now raises `ValueError` when credentials are present but the hostname is missing.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: devinfra/claude/testing/proxy_ca.py [hunk 14]
```diff
@@ -14,6 +14,35 @@
...
+        x509.CertificateBuilder()
+        .subject_name(subject)
+        .issuer_name(issuer)
+        .public_key(key.public_key())
+        .serial_number(x509.random_serial_number())
+        .not_valid_before(datetime.now(UTC))
+        .not_valid_after(datetime.now(UTC) + timedelta(days=365))
+        .add_extension(x509.BasicConstraints(ca=False, path_length=None), critical=True)
+        .add_extension(x509.SubjectAlternativeName([x509.IPAddress(ipaddress.ip_address(host))]), critical=False)
+        .sign(key, hashes.SHA256())
```
[the remark this answers is not in the document]
Comment (agentydragon) [reply devinfra/claude/testing/proxy_ca.py:43]: Fixed. `generate_server_cert()` now falls back to `x509.DNSName(host)` when the host is not a valid IP address.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: devinfra/claude/hook_daemon/test_tracing.py (hunk 8 above)
Comment (agentydragon) [reply devinfra/claude/hook_daemon/test_tracing.py:55]: Fixed. `test_configure_no_endpoint_is_noop` now uses the `session` fixture.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: devinfra/claude/hook_daemon/test_tracing.py (hunk 9 above)
Comment (agentydragon) [reply devinfra/claude/hook_daemon/test_tracing.py]: Fixed. Added a `span` fixture returning `ReadableSpan(name="test-span")`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: devinfra/claude/hook_daemon/session_start/test_k8s_proxy_integration.py [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    cert_file = tmp_path / "server.crt"
+    key_file = tmp_path / "server.key"
+    cert_file.write_bytes(cert_pem)
+    key_file.write_bytes(key_pem)
+
+    tls_ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
+    tls_ctx.load_cert_chain(str(cert_file), str(key_file))
+
+    server = http.server.HTTPServer(("0.0.0.0", port), _K8sHandler)
+    server.socket = tls_ctx.wrap_socket(server.socket, server_side=True)
```
[the remark this answers is not in the document]
Comment (agentydragon) [reply devinfra/claude/hook_daemon/session_start/test_k8s_proxy_integration.py:78]: Fixed. Added `tls_ctx.minimum_version = ssl.TLSVersion.TLSv1_2` after creating the `SSLContext`.

---
_Generated by [Claude Code](https://claude.ai/code)_
